dispatches/models/renewables_case/wind_battery_PEM_tank_turbine_LMP.py

```diff
@@ -0,0 +1,516 @@
+    m.fs.windpower.setup_resource()
+
+    outlvl = idaeslog.INFO if verbose else idaeslog.WARNING
+    m.fs.windpower.initialize(outlvl=outlvl)
+    propagate_state(m.fs.wind_to_splitter)
+    m.fs.splitter.initialize()
+    return m
+
+
+def wind_battery_pem_tank_turb_optimize(n_time_points, h2_price=h2_price_per_kg, tank_type="simple", verbose=False, plot=False):
+    from timeit import default_timer
+    start = default_timer()
+    # create the multiperiod model object
+    mp_model = MultiPeriodModel(n_time_points=n_time_points,
+                                process_model_func=partial(wind_battery_pem_tank_turb_mp_block, tank_type=tank_type, verbose=verbose),
+                                linking_variable_func=wind_battery_pem_tank_turb_variable_pairs,
+                                periodic_variable_func=wind_battery_pem_tank_turb_periodic_variable_pairs)
+
+    mp_model.build_multi_period_model(wind_resource)
+
+    m = mp_model.pyomo_model
+    blks = mp_model.get_active_process_blocks()
+
+    using_simple_tank = "Simple" in type(blks[0].fs.h2_tank).__name__
+    if not using_simple_tank and use_simple_h2_tank:
+        # turn off energy holdup constraints
+        for blk in blks:
+            if hasattr(blk, "link_constraints"):
+                blk.link_constraints[1].deactivate()
+            if hasattr(blk, "periodic_constraints"):
+                blk.periodic_constraints[1].deactivate()
+
+    m.h2_price_per_kg = pyo.Param(default=h2_price, mutable=True)
+    m.pem_system_capacity = Var(domain=NonNegativeReals, initialize=fixed_pem_mw * 1e3, units=pyunits.kW)
+    m.h2_tank_size = Var(domain=NonNegativeReals, initialize=fixed_tank_size)
+    m.turb_system_capacity = Var(domain=NonNegativeReals, initialize=turb_p_mw * 1e3, units=pyunits.kW)
+    if not design_opt:
+        m.pem_system_capacity.fix(fixed_pem_mw * 1e3)
+        m.h2_tank_size.fix(fixed_tank_size)
+        m.turb_system_capacity.fix(turb_p_mw * 1e3)
+        m.contract_capacity = Var(domain=NonNegativeReals, initialize=0, units=pyunits.mol / pyunits.second)
+
+    m.wind_cap_cost = pyo.Param(default=wind_cap_cost, mutable=True)
+    if extant_wind:
+        m.wind_cap_cost.set_value(0.)
+    m.pem_cap_cost = pyo.Param(default=pem_cap_cost, mutable=True)
+    m.batt_cap_cost = pyo.Param(default=batt_cap_cost, mutable=True)
+    m.tank_cap_cost = pyo.Param(default=tank_cap_cost_per_kg, mutable=True)
+    m.turb_cap_cost = pyo.Param(default=turbine_cap_cost, mutable=True)
+
+    for blk in blks:
+        blk_wind = blk.fs.windpower
+        blk_battery = blk.fs.battery
+        blk_pem = blk.fs.pem
+        blk_tank = blk.fs.h2_tank
+        blk_turb = blk.fs.h2_turbine
+        # add operating constraints
+        blk_pem.max_p = Constraint(blk_pem.flowsheet().config.time,
+                                   rule=lambda b, t: b.electricity[t] <= m.pem_system_capacity)
```

Comment:
This works and there is no need to change it. I just wanted to note that this not a recommended practice. In general, variables from a parent block (in this case `m.pem_system_capacity`) are not recommended to be used in its child block (`blk_pem` in this case). This is because it can create issues especially if you want to solve the child block separately. To my knowledge, this is the recommended approach. Instead of adding the constraint to `blk_pem` object, add it to `m` i.e., the parent block which contains the variable `m.pem_system_capacity`.

Reply:
thanks I've moved all these constraints to the parent block:

```
    m.pem_max_p = Constraint(mp_wind_pem.pyomo_model.TIME,
                                rule=lambda b, t: blks[t].fs.pem.electricity[0] <= m.pem_system_capacity)
```